### ISSUE TITLE
TSDAE cuts off last input to decoder

### ISSUE THREAD

**Issue by zynaa** [UKPLab/sentence-transformers#1676]
Hey, thanks for the awesome work here!

I was looking at the TSADE implementation and noticed the following in the loss function in `losses\DenoisingAutoEncoderLoss.py`: 

```python
# Prepare input and output
target_length = target_features['input_ids'].shape[1]
decoder_input_ids = target_features['input_ids'].clone()[:, :target_length - 1]
label_ids = target_features['input_ids'][:, 1:]

# Decode
decoder_outputs = self.decoder(
    input_ids=decoder_input_ids,
    inputs_embeds=None,
    attention_mask=None,
    encoder_hidden_states=reps[:, None],  # (bsz, hdim) -> (bsz, 1, hdim)
    encoder_attention_mask=source_features['attention_mask'][:, 0:1],
    labels=None,
    return_dict=None,
    use_cache=False
)

# Calculate loss
lm_logits = decoder_outputs[0]
ce_loss_fct = nn.CrossEntropyLoss(ignore_index=self.tokenizer_decoder.pad_token_id)
loss = ce_loss_fct(lm_logits.view(-1, lm_logits.shape[-1]), label_ids.reshape(-1))
return loss
```

My understanding was, that for causal language modelling the shifting needs to be applied on the logits instead of the input ids , like e.g. in `BertLMHeadModel`:

```python
shifted_prediction_scores = prediction_scores[:, :-1, :].contiguous()
labels = labels[:, 1:].contiguous()
loss_fct = CrossEntropyLoss()
lm_loss = loss_fct(shifted_prediction_scores.view(-1, self.config.vocab_size), labels.view(-1))
```

So I would have expected the loss function for TSADE to look like this instead:

```python
# Prepare input and output
label_ids = target_features['input_ids'].clone()
label_ids = label_ids[:, 1:]

# Decode
decoder_outputs = self.decoder(
    input_ids=target_features['input_ids'],
    inputs_embeds=None,
    attention_mask=target_features['attention_mask'],
    encoder_hidden_states=reps[:, None],  # (bsz, hdim) -> (bsz, 1, hdim)
    encoder_attention_mask=source_features['attention_mask'][:, 0:1],
    labels=None,
    return_dict=None,
    use_cache=False
)

# Calculate loss
lm_logits = decoder_outputs[0]
lm_logits = lm_logits[:, :-1]
ce_loss_fct = nn.CrossEntropyLoss(ignore_index=self.tokenizer_decoder.pad_token_id)
loss = ce_loss_fct(lm_logits.reshape(-1, lm_logits.shape[-1]), label_ids.reshape(-1))
return loss
```
Is the current behaviour intended, or is this a bug?

As a side-note: the code snippet above could possibly also be very slightly improved by also avoiding the calculation of the loss on other special tokens with the use of the special tokens mask returned by the tokenizer (instead of just saying `ignore_index=self.tokenizer_decoder.pad_token_id`)

**Comment by zynaa** [comment]
Okay nevermind, it should already be correct like this, since the last token is always masked anyways.
